Introduce missing team role if necessary (#23243)

raboof opened the pull request:
When receiving gossip from a node that did not contain any team
information (such as gossip from a node running a previous version of
Akka), add the default team role during deserialization.

Refs #23243

raboof committed:
Introduce cluster 'team' setting and add to Member

Introduced cluster-team.md so we can grow the documentation with each
PR, but did not add it to the ToC yet.

(cherry picked from commit a06badaa03fa9f3c9a942b1468090f758c74a869)

raboof committed:
Less abbreviations, more reliable test

(cherry picked from commit 61e289b276f410654c1b063c33648e0d7ea88e50)

raboof committed:
Introduce missing team role if necessary (#23243)

When receiving gossip from a node that did not contain any team
information (such as gossip from a node running a previous version of
Akka), add the default team role during deserialization.

ktoso commented:
Refs #23243

patriknw commented on akka-cluster/src/main/scala/akka/cluster/protobuf/ClusterMessageSerializer.scala:
```diff
@@ -336,7 +336,16 @@ class ClusterMessageSerializer(val system: ExtendedActorSystem) extends BaseSeri
+        val headrole = roleMapping(head)
+        if (headrole.startsWith("team-")) resultSoFar ++ roleIndexes.map(roleMapping(_)).toSet
+        else rolesFromProto(tail, resultSoFar + headrole)
+    }
```
looks somewhat expensive. Not sure what collection type the `getRolesIndexesList.asScala` is. Perhaps we could iterate the raw java List and build the `Set` in a more efficient way than `++`?

Hawstein replied:
Indeed. `getRolesIndexesList.asScala` is a `JListWrapper` which extends `mutable.Buffer`. `tail` costs `O(n)` instead of `O(1)`

patriknw reviewed (changes requested):
good solution, but perhaps perf could be improved

Hawstein commented on akka-cluster/src/main/scala/akka/cluster/protobuf/ClusterMessageSerializer.scala:
```diff
@@ -336,7 +336,16 @@ class ClusterMessageSerializer(val system: ExtendedActorSystem) extends BaseSeri
+    @tailrec
+    def rolesFromProto(roleIndexes: Seq[Integer], resultSoFar: Set[String] = Set.empty): Set[String] = roleIndexes match {
+      case Seq() ⇒ resultSoFar + "team-default"
+      case Seq(head, tail @ _*) ⇒
```
Why not here just use:

```
  case Nil => blabla
  case head :: tail => blabla
```

Hawstein replied:
Oh, i know. Because at the call site, this method is given a `mutable.Buffer[Integer]`.

patriknw committed:
Revert "Less abbreviations, more reliable test"

This reverts commit 61e289b276f410654c1b063c33648e0d7ea88e50.

patriknw committed:
Revert "Introduce cluster 'team' setting and add to Member"

This reverts commit a06badaa03fa9f3c9a942b1468090f758c74a869.

raboof committed:
Simpler implementation of adding default role

Though the benchmark didn't really reliably show a difference between this and
only roleIndexes.map(roleMapping(_)).toSet in the first place, so perhaps it
doesn't really matter. Extra reason to choose the more readable version.

raboof commented:
Superseded by #23276
